File: src/main/java/nl/waterjeloen/quoridor/BoardPanel.java
package nl.waterjeloen.quoridor;

import javax.swing.JComponent;
import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Rectangle;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class BoardPanel extends JComponent {
    private static final Color[] PLAYER_COLORS = { Color.RED, Color.BLUE, Color.GREEN, Color.YELLOW };

    private final Board board;
    private final List<BoardListener> listeners;
    private final List<Rectangle> highlights;

    public BoardPanel(Board board) {
        this.board = board;
        this.listeners = new ArrayList<>();
        this.highlights = new ArrayList<>();

        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent event) {
                final int x = event.getX();
                final int y = event.getY();

                final Rectangle cell = calculateCell();
                if ((x < cell.x) || (x > cell.x + (board.getSize() * 7 - 1) * cell.width)
                    || (y < cell.y) || (y > cell.y + (board.getSize() * 7 - 1) * cell.height)) {
                    return; // outside board
                }

                final int cellX = (x - cell.x) / cell.width;
                final int cellY = (y - cell.y) / cell.height;
                final Location location = new Location(cellY / 7, cellX / 7);

                final boolean isHorizontalWall = cellY % 7 == 6;
                final boolean isVerticalWall = cellX % 7 == 6;
                for (BoardListener l : listeners) {
                    if (!isHorizontalWall && !isVerticalWall) {
                        l.fieldClicked(location);
                    } else if (isHorizontalWall && isVerticalWall) {
                        l.centerClicked(location);
                    } else if (isHorizontalWall) {
                        l.horizontalWallClicked(location);
                    } else { // (isVerticalWall)
                        l.verticalWallClicked(location);
                    }
                }
            }
        });
    }

    public void addListener(BoardListener listener) {
        listeners.add(listener);
    }

    public void highlightField(Location location) {
        if (location.isValid(board.getSize())) {
            highlights.add(new Rectangle(location.column * 7, location.row * 7, 6, 6));
            repaint();
        }
    }

    public void highlightCenter(Location location) {
        if (location.isValid(board.getSize() - 1)) {
            highlights.add(new Rectangle(location.column * 7 + 6, location.row * 7 + 6, 1, 1));
            repaint();
        }
    }

    public void highlightHorizontalWall(Location location) {
        if (location.isValid(board.getSize() - 1, board.getSize())) {
            highlights.add(new Rectangle(location.column * 7, location.row * 7 + 6, 6, 1));
            repaint();
        }
    }

    public void highlightVerticalWall(Location location) {
        if (location.isValid(board.getSize(), board.getSize() - 1)) {
            highlights.add(new Rectangle(location.column * 7 + 6, location.row * 7, 1, 6));
            repaint();
        }
    }

    public void clearHighlights() {
        highlights.clear();
        repaint();
    }

    @Override
    public void paintComponent(Graphics graphics) {
        final Graphics2D g = (Graphics2D) graphics;

        final Rectangle cell = calculateCell();

        final Set<Location> reachable = board.getReachableLocations(board.getCurrentPlayer().getLocation());
        for (int r = 0; r < board.getSize(); ++r) {
            for (int c = 0; c < board.getSize(); ++c) {
                final Location location = new Location(r, c);
                final Color color = (location.equals(board.getCurrentPlayer().getLocation())) ? Color.GRAY.darker()
                        : (board.getCurrentPlayer().winsAt(location) && reachable.contains(location)) ? Color.GRAY.darker()
                        : Color.BLACK;
                g.setColor(color);
                g.fillRect(
                    cell.x + cell.width * c * 7,
                    cell.y + cell.height * r * 7,
                    cell.width * 6,
                    cell.height * 6);
            }
        }

        g.setColor(Color.ORANGE);
        for (int r = 0; r < board.getSize() - 1; ++r) {
            for (int c = 0; c < board.getSize() - 1; ++c) {
                final Location location = new Location(r, c);
                if (board.hasHorizontalWall(location)) {
                    g.fillRect(
                        cell.x + cell.width * c * 7,
                        cell.y + cell.height * (r * 7 + 6),
                        cell.width * 13,
                        cell.height);
                }
                if (board.hasVerticalWall(location)) {
                    g.fillRect(
                        cell.x + cell.width * (c * 7 + 6),
                        cell.y + cell.height * r * 7,
                        cell.width,
                        cell.height * 13);
                }
            }
        }
        g.setColor(Color.GREEN);
        for (Rectangle r : highlights) {
            g.fillRect(
                cell.x + cell.width * r.x,
                cell.y + cell.height * r.y,
                cell.width * r.width,
                cell.height * r.height);
        }
        for (int i = 0; i < board.getPlayerCount(); ++i) {
            final Player player = board.getPlayer(i);
            g.setColor(PLAYER_COLORS[i]);
            g.fillOval(
            cell.x + cell.width * (player.getLocation().column * 7 + 1),
            cell.y + cell.height * (player.getLocation().row * 7 + 1),
            cell.width * 4,
            cell.height * 4);
            g.setColor(Color.WHITE);
            g.drawString(Integer.toString(player.getWalls()), cell.x + cell.width * (player.getLocation().column * 7 + 1), cell.y + cell.height * (player.getLocation().row * 7 + 1));
        }
    }

    private Rectangle calculateCell() {
        final int cellWidth = getWidth() / (board.getSize() * 7 - 1);
        final int cellHeight = getHeight() / (board.getSize() * 7 - 1);
        final int cellSize =  Math.min(cellWidth, cellHeight);
        final int xOffset = (getWidth() - cellSize * (board.getSize() * 7 - 1)) / 2;
        final int yOffset = (getHeight() - cellSize * (board.getSize() * 7 - 1)) / 2;
        return new Rectangle(xOffset, yOffset, cellSize, cellSize);
    }
}

File: src/main/scala/nl/zwaaltjes/quoridor/BoardPanel.scala
package nl.zwaaltjes.quoridor

import java.awt.{Color, Font, Graphics, Graphics2D, Rectangle, RenderingHints, Toolkit}
import java.awt.event.{MouseAdapter, MouseEvent}
import javax.swing.JComponent

object BoardPanel {
  private val playerColors = Array(Color.red, Color.blue, Color.green, Color.yellow)

  private val side = 7
}

class BoardPanel(board: Board) extends JComponent {
  import BoardPanel.*

  private val colorMap = {
    val entries = for (i <- 0 until board.playerCount) yield board.player(i) -> playerColors(i)
    entries.toMap
  }
  private var listeners = List.empty[Board.Listener]
  private var highlights = List.empty[Rectangle]

  addMouseListener(new MouseAdapter {
    override def mouseClicked(event: MouseEvent): Unit = {
      val x = event.getX
      val y = event.getY
      val cell = cellDimension
      val max = board.size * side - 1
      if ((x >= cell.x) && (x <= cell.x + max * cell.width) && (y >= cell.y) && (y <= cell.y + max * cell.height)) {
        val cellX = (x - cell.x) / cell.width
        val cellY = (cell.y + max * cell.height - y) / cell.height
        val location = Location(cellY / side, cellX / side)
        val isHorizontalWall = cellY % side == (side - 1)
        val isVerticalWall = cellX % side == (side - 1)
        listeners.foreach { l =>
          if (!isHorizontalWall && !isVerticalWall)
            l.fieldClicked(location)
          else if (isHorizontalWall && isVerticalWall)
            l.centerClicked(location)
          else if (isHorizontalWall)
            l.horizontalWallClicked(location)
          else // (isVerticalWall)
            l.verticalWallClicked(location)
        }
      }
    }
  })

  def addListener(listener: Board.Listener): Unit =
    listeners ::= listener

  private def xOffset(column: Int): Int = column * side
  private def yOffset(row: Int): Int = (board.size - row - 1) * side

  def highlightField(location: Location): Unit =
    if (location.isValid(board.size)) {
      highlights ::= new Rectangle(xOffset(location.column), yOffset(location.row), side - 1, side - 1)
      repaint()
    }

  def highlightCenter(location: Location): Unit =
    if (location.isValid(board.size - 1)) {
      highlights ::= new Rectangle(xOffset(location.column + 1) - 1, yOffset(location.row) - 1, 1, 1)
      repaint()
    }

  def highlightHorizontalWall(location: Location): Unit =
    if (location.isValid(board.size - 1, board.size)) {
      highlights ::= new Rectangle(xOffset(location.column), yOffset(location.row) - 1, side - 1, 1)
      repaint()
    }

  def highlightVerticalWall(location: Location): Unit =
    if (location.isValid(board.size, board.size - 1)) {
      highlights ::= new Rectangle(xOffset(location.column + 1) - 1, yOffset(location.row), 1, side - 1)
      repaint()
    }

  def clearHighlights(): Unit = {
    highlights = Nil
    repaint()
  }

  override def paintComponent(graphics: Graphics): Unit = {
    val g = graphics.asInstanceOf[Graphics2D]
    g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
    g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON)
    val cell = cellDimension

    def xOf(column: Int): Int = cell.x + column * cell.width * side
    def yOf(row: Int): Int = cell.y + (board.size - row - 1) * cell.height * side

    // draw indices
    g.setColor(Color.black)
    for (r <- 0 until board.size) {
      val text = Location.rowString(r)
      drawText(g, text, new Rectangle(xOf(-1), yOf(r), cell.width, cell.height))
    }
    for (c <- 0 until board.size) {
      val text = Location.columnString(c)
      drawText(g, text, new Rectangle(xOf(c), yOf(-1), cell.width, cell.height))
    }

    // draw fields
    val reachable = board.reachableLocations(board.currentPlayer.location)
    for {
      r <- 0 until board.size
      c <- 0 until board.size
      location = Location(r, c)
    } {
      val color = if (board.winner.isDefined) Color.black
        else if (location == board.currentPlayer.location) Color.gray.darker
        else if (board.currentPlayer.wins(location) && reachable.contains(location)) colorMap(board.currentPlayer).darker.darker
        else Color.black
      g.setColor(color)
      g.fillRect(xOf(c), yOf(r), cell.width * (side - 1), cell.height * (side - 1))
    }

    // draw walls
    g.setColor(Color.orange.darker)
    for {
      r <- 0 until board.size - 1
      c <- 0 until board.size - 1
      location = Location(r, c)
    } {
      if (board.hasHorizontalWall(location))
        g.fillRect(xOf(c), yOf(r) - cell.height, cell.width * (2 * side - 1), cell.height)
      if (board.hasVerticalWall(location))
        g.fillRect(xOf(c + 1) - cell.width, yOf(r + 1), cell.width, cell.height * (2 * side - 1))
    }

    g.setColor(Color.orange)

    // draw highlights
    for (r <- highlights) {
      g.fillRect(cell.x + cell.width * r.x, cell.y + cell.height * r.y, cell.width * r.width, cell.height * r.height)
    }

    // draw players
    for (i <- 0 until board.playerCount) {
      val player = board.player(i)
      g.setColor(playerColors(i).darker)
      g.fillOval(
        xOf(player.location.column) + cell.width,
        yOf(player.location.row) + cell.height,
        cell.width * (side + 1) / 2,
        cell.height * (side + 1) / 2,
      )
      g.setColor(Color.white)
      drawText(g, player.walls.toString, new Rectangle(
        xOf(player.location.column),
        yOf(player.location.row),
        cell.width / 3,
        cell.height / 3,
      ))
    }
  }

  private def drawText(graphics: Graphics2D, text: String, box: Rectangle): Unit = {
    val fontSize = 72f * box.height * side / Toolkit.getDefaultToolkit.getScreenResolution
    graphics.setFont(new Font(Font.SERIF, Font.PLAIN, fontSize.toInt))
    val metrics = graphics.getFontMetrics

    val bounds = metrics.getStringBounds(text, graphics)
    val offset = (box.width * side - bounds.getWidth.toInt) / 2
    graphics.drawString(text, box.x + offset, box.y + metrics.getAscent)
  }

  private def cellDimension: Rectangle = {
    val cellWidth = getWidth / ((board.size + 1) * side - 1)
    val cellHeight = getHeight / ((board.size + 1) * side - 1)
    val cellSize = cellWidth min cellHeight
    val xOffset = (getWidth - cellSize * ((board.size + 1) * side - 1)) / 2
    val yOffset = (getHeight - cellSize * ((board.size + 1) * side - 1)) / 2
    new Rectangle(xOffset + cellSize * side, yOffset, cellSize, cellSize)
  }
}